lib/ansible/plugins/strategy/__init__.py

```diff
@@ -481,6 +481,8 @@ def parent_handler_match(target_handler, handler_name):
                                 if original_task.action == 'set_fact':
                                     self._variable_manager.set_nonpersistent_facts(target_host, result_item['ansible_facts'].copy())
                                 else:
+                                    if original_task.action == 'setup':
+                                        target_host._gathered_facts = True
```

Comment:
'before' it was checked directly in play_iterator (before that in runner), if the host property is only used there, that would remove the need for it. 
The other advantage is that it also prevents gathering if the 'cache' is fresh.